src/WebJobs.Script/Host/HostIdValidator.cs

```diff
@@ -0,0 +1,198 @@
+                    // Schedule the validation to run asynchronously after a delay. This delay ensures
+                    // we're not impacting coldstart, and also gives time for the primary host to be
+                    // identified.
+                    Utility.ExecuteAfterDelay(() => Task.Run(() => ValidateHostIdUsageAsync(hostId)), TimeSpan.FromSeconds(validationDelaySeconds));
+                    _validationScheduled = true;
+                }
+            }
+        }
+
+        internal async Task ValidateHostIdUsageAsync(string hostId)
+        {
+            try
+            {
+                if (!_storageProvider.ConnectionExists(ConnectionStringNames.Storage))
+                {
+                    return;
+                }
+
+                var primaryHostStateProvider = _serviceProvider.GetScriptHostServiceOrNull<IPrimaryHostStateProvider>();
+                if (primaryHostStateProvider != null && !primaryHostStateProvider.IsPrimary)
+                {
+                    // as an optimization, only validate on the primary host
+                    return;
+                }
+
+                HostIdInfo hostIdInfo = await ReadHostIdInfoAsync(hostId);
+
+                if (hostIdInfo != null)
+                {
+                    // an existing record exists for this host ID
+                    // verify the host name is the same as our host name
+                    if (string.Compare(_hostNameProvider.Value, hostIdInfo.Hostname, StringComparison.OrdinalIgnoreCase) != 0)
+                    {
+                        HandleCollision(hostId);
+                    }
+                }
+                else
+                {
+                    // no existing record, so write one, claiming this host ID for this host name
+                    // in this storage account
+                    hostIdInfo = new HostIdInfo
+                    {
+                        Hostname = _hostNameProvider.Value
+                    };
+                    await WriteHostIdAsync(hostId, hostIdInfo);
+                }
+            }
+            catch (Exception ex)
+            {
+                // best effort - log error and continue
+                _logger.LogError(ex, "Error validating host ID usage.");
+            }
+        }
+
+        private void HandleCollision(string hostId)
+        {
+            // see if the user has specified a level, otherwise default
+            LogLevel level = DefaultLevel;
+            string value = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsHostIdCheckLevel);
+            if (value != null && Enum.TryParse<LogLevel>(value, out LogLevel parsedLevel))
+            {
+                level = parsedLevel;
+            }
+
+            string message = string.Format(Resources.HostIdCollisionFormat, hostId);
+            if (level == LogLevel.Error)
+            {
+                _logger.LogError(message);
+                _applicationLifetime.StopApplication();
```

Comment:
Just curious about the timing for this. Is it possible that some listeners would have started and some function executions are in progress by the time this StopApplication call is made? (I get that we don't want to block HTTP executions on this check to avoid cold start impact)

Reply:
Yes - to avoid coldstart impact we don't block startup on the check. So we're optimizing for apps with no issues - if an app does have a collision then yes, the hard failure will occur 10s after startup.

Reply:
Got it. I am a little worried about side effects, like a large number of messages being written to poison queues, because the app is in a restart loop due to this error. For HTTP there is nothing we can/should do because we dont want to impact cold start. But for retriable event sources like queues, the impact of terminating the app 10 seconds after start could be a problem. I think it might be better to block listeners on this check, though not sure if that is going to be straightforward..

Reply:
Http processing for the host is gated on all hosted services starting. The hosted service WebJobsScriptHostService is responsible for starting the job host (and starting the listeners) and the host is started [here](https://github.com/Azure/azure-functions-host/blob/4764c25d885cdfb1d84af3b80861b181ce658faa/src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs#L292). If we were to change that service to only start the JobHost _after_ the host ID check it would ensure we only start listeners after the ID is validated, however all apps would then be paying a coldstart cost (both for http, as well as background triggers since their startup is delayed). Given the lengths we've went to recently to remove all such storage calls off the startup path, I don't think we want to regress this.

A compromise/improvement I can make is to remove the IsPrimary check which will allow me to reduce the delay on this check to our standard coldstart delay of 5s, which minimizes how long the host is allowed to run before the issue is discovered. That does mean all instances will be doing the check, but that's not a huge problem.

Reply:
yeah I agree this can't be part of the http startup path.

For now I'm going to assume we just don't have the code structured in a way that can allow a sequence where 1) http listener starts, 2) checks run, 3) other listeners start.. I get that we only have a limited amount of time to put this check in and we can't be engaging in larger refactorings right now..

I do not think the 5 seconds vs 10 seconds will make much of a difference. Either way its enough time for us to pull a bunch of messages of queues, start processing user code, and then terminate. 

Even with this potential problem, I still think the change is worth it. Just wanted to make sure we talked it through.

Reply:
Actually wait a minute..your comment about IsPrimary makes me wonder.. why are we checking that? Lets say the customer has configured the app to run on an app service plan with 3 instances.. are you saying the primary will shutdown, but the other two instances will run?

Reply:
Ha good point - yeah you're right :) IsPrimary made sense when this was just an error log, but once we moved to host shutdown, clearly we need that to run on all instances. I'll remove the IsPrimary check.